### PR TITLE
Use Task to build the project

### PR DESCRIPTION
**Pull request by wraymo**
# Description

The package is currently built using a Python script (the so-called "packager") which has no support for incremental builds. In addition, the way we install the Python components into the package is quite hacky and doesn't take advantage of Python's packaging tools.

This PR adds support for building the Python components using [poetry](https://python-poetry.org/) and replaces the packager tool with a Taskfile (see [Task](https://taskfile.dev/)). The Taskfile allows us to track changes to source files and only rebuild artifacts that depend on the changed source files.

One limitation of this PR (compared to the original packager) is that we haven't added support for automatically building the package in a container. As a result, users of this new build flow will need to setup a build machine/container manually. We hope to resolve this in a future PR.

# Validation performed

- Passed yamllint
- Successfully built the package using Task
- Validated compression and search with the [hive-24hrs](https://zenodo.org/records/7094921#.Y5JbH33MKHs) dataset.

### CODE REVIEW

File: BUILDING.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Requirements
+
+* [Meteor](https://docs.meteor.com/install.html#installation)
+* [NodeJS 14](https://nodejs.org/download/release/v14.21.3/)
```
Comment (kirkrodrigues) [inline BUILDING.md:4]: Meteor and NodeJS aren't requirements right now.

File: BUILDING.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+* python3-venv
+* [Task](https://taskfile.dev/)
+
+# Building a package
+
+* Run `task` from the root of the project
+* The package will be available at `out/clp-package`
+
+NOTE: The taskfile is not set up perfectly so it's possible it may generate an
```
Comment (kirkrodrigues) [inline BUILDING.md:13]: We should address this limitation in this PR.

File: BUILDING.md
Comment (kirkrodrigues) [inline BUILDING.md:1]: This doc should be in `docs/Building.md`.

File: components/clp-py-utils/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+mysql-connector-python = "~8.0.26"
+pydantic = "^1.8.2"
+python-Levenshtein = "^0.21.1"
+PyYAML = "^6.0.1"
+
+[tool.poetry.group.dev.dependencies]
+black = "^23.9.1"
+docformatter = "^1.7.5"
+mypy = "^1.5.1"
+ruff = "^0.0.291"
```
Comment (kirkrodrigues) [inline components/clp-py-utils/pyproject.toml:20]: Let's add these dev dependencies in another PR.

File: requirements.txt
Comment (kirkrodrigues) [inline requirements.txt:1]: Do we need these? Shouldn't this all be handled in the taskfile?

File: components/compression-job-handler/pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+[tool.poetry.group.dev.dependencies]
+black = "^23.9.1"
+docformatter = "^1.7.5"
+mypy = "^1.5.1"
+ruff = "^0.0.291"
+
+[build-system]
+requires = ["poetry-core"]
+build-backend = "poetry.core.masonry.api"
```
Comment (kirkrodrigues) [inline components/compression-job-handler/pyproject.toml:25]: Missing empty line at end of file.

File: components/clp-py-utils/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+[tool.poetry]
+name = "clp-py-utils"
+version = "0.1.0"
+description = "Utilities for other Python packages in CLP."
+authors = ["YScope Inc. <dev@yscope.com>"]
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = "^3.8 || ^3.10"
+mariadb = "~1.0.7"
```
Comment (kirkrodrigues) [inline components/clp-py-utils/pyproject.toml:10]: Did you check whether each component needs all these dependencies? E.g., does `clp-py-utils` require `mariadb` and `mysql-connector-python`?

File: Taskfile.yml [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+  clean:
+    cmds:
+      - rm -rf {{.CLP_BUILD_DIR}}
+
+  clean_package:
+    cmds:
+      - rm -rf {{.CLP_BUILD_DIR}}/clp-package
+
+  # TODO Clean-up
```
Comment (kirkrodrigues) [inline Taskfile.yml:20]: Should get rid of the TODO.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
A quick review (we will need another iteration after this). Some high-level comments as well:
* You're missing the validation section of the PR?
* You should add the build folder of each component to `.gitignore`
* The taskfile replaces the `packager` so we should ensure they generate the same output (from a user's perspective) and that we remove unnecessary files/docs, etc.

File: components/clp-py-utils/pyproject.toml (hunk 5 above)
Comment (wraymo) [reply components/clp-py-utils/pyproject.toml:10]: Yeah, sql_adapter.py needs these two

File: components/clp-py-utils/pyproject.toml (hunk 3 above)
Comment (wraymo) [reply components/clp-py-utils/pyproject.toml:20]: Ok

File: requirements.txt
Comment (wraymo) [reply requirements.txt:1]: We need poetry to build Python packages.

Comment (kirkrodrigues) [reply requirements.txt:1]: Ah right, sorry, I forgot how the Taskfile was using this.

File: components/clp-package-utils/README.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# CLP Package Utilities
+
+This python module contains utilities used by thegit  CLP package.
```
Comment (kirkrodrigues) [inline components/clp-package-utils/README.md:3]:
```suggestion
This python module contains utilities used by the CLP package.
```

File: requirements.txt
Comment (kirkrodrigues) [inline requirements.txt:1]: To ensure reproducible builds, let's lock the versions of these packages.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
I have changes locally that I will commit once I clean them up, but in the mean time:
* You're still missing the validation section of the PR?
* For the Python components, we don't need the `constraints.txt` and `requirements.txt` files since we're now using `pyproject.toml`
* Similarly, in the Python component READMEs, the instructions should be about how to build and install with `poetry`.

**Review by wraymo — COMMENTED** [review]
Looks good to me

**Review by kirkrodrigues — APPROVED** [review]

**Pull request merged**